src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_search.cs

```diff
@@ -471,19 +473,15 @@ private void SkipPastAnySlow(ReadOnlySpan<T> values)
         public bool IsNext(ReadOnlySpan<T> next, bool advancePast)
         {
             ReadOnlySpan<T> unread = UnreadSpan;
-            if (unread.Length >= next.Length)
+            if (unread.StartsWith(next))
             {
-                if (unread.StartsWith(next))
+                if (advancePast)
                 {
-                    if (advancePast)
-                    {
-                        Advance(next.Length);
-                    }
-                    return true;
+                    Advance(next.Length);
                 }
-                return false;
+                return true;
             }
-            return IsNextSlow(next, advancePast);
+            return unread.Length < next.Length && IsNextSlow(next, advancePast);
```

Comment:
This is relying on boolean expression short-circuiting.

`if (unread.Length >= next.Length)` is true, `unread.Length < next.Length` is false.

if `unread.Length < next.Length` is false, we return false, and do not call `IsNextSlow()`.

if `unread.Length < next.Length` is true, `IsNextSlow()` decides whether we return true or false.